### DISCUSSION TITLE
Add permission nodes to permit specific rgb colors in chat

### DISCUSSION THREAD

**Discussion by accLarson** [EssentialsX/Essentials#5224] [Ideas and Feature Suggestions] [answered]
Consider adding a permission node which permits a player to use only the specified hex code in chat.

Currently, admins can assign permissions as follows:
`essentials.chat.aqua`
`essentials.chat.red`
etc
`essentials.chat.rgb` for all HEX colors

I propose a new permission node:

`essentials.chat.#000000` or `essentials.chat.000000` for allowing the use of specific hex color codes.

This system should probably also be extended to other permission nodes that have similar systems (essentials.msg, essentials.nick, essentials.signs)

**Comment by pop4959** [comment, marked as answer]
Can you elaborate on a use case for such a permission?

**Reply by accLarson** [reply]
For when an admin desires to allow some cases of hex color use but not ALL hex color use.

In my case, I'd like to actually use Player#chat in a plugin and have the player's message display with color. This seems to no longer be possible within the last year or so. If I grant the essentials.chat.rgb permission, I can prefix their message with the color code I want however this would give them permission to use all color codes if they found out the essentials formatting (&#000000)

This use to be possible by just adding a color to Player#chat but has recently been changed https://github.com/PaperMC/Paper/issues/8379

To clarify, the plugin I'm writting is for /distinguish which would send the with a elevated/noticeable color rather than their default chat color.

**Reply by pop4959** [reply]
Using Essentials for this sounds like a strange workaround then. Is there a reason why it isn't handled in your plugin?

**Reply by accLarson** [reply]
Agreed, it's not an ideal solution but I figured the permission nodes might be justifiable enough irrelevant to my obscure use case.

I'm hesitant to turn this discussion into a help session for my own plugin development but since you asked, I'll explain.

My previous method of handling this within my plugin is no longer possible with recent changes to `Player#chat` and how it handles color being passed in. https://github.com/PaperMC/Paper/issues/8379. Previously, I could concatenate a `ChatColor#of` or `TextColor#fromHexString` to my argument being passed into `Player#chat`. Now this kicks the player for `Illegal characters in chat`.

This issue is specific to my need to send this message FROM the player (`Player#chat`). I could instead `Server#broadcast` a similar message using some format like `<username>:<color><message>` however this loses all the other formatting on the player like the clan, prefix, suffix, etc from various other plugins. My end goal was to just allow players to distinguish their chat with a bold hex color.

**Reply by accLarson** [reply]
It just occurred to me that Essentials must have some method of doing what I'm trying to do since Essentials can be configured to have players send chat messages with hex colors. (withing group formatting and also just using the permission node `essentials.chat.rgb`)

**Reply by accLarson** [reply]
Likely not using `Player#chat` I assume. Probably using the `AsyncPlayerChatEvent`?

**Reply by pop4959** [reply]
Yeah, ideally you want to avoid using chat color codes directly. You might want to look into stuff like [ChatRenderer](https://jd.papermc.io/paper/1.19/io/papermc/paper/chat/ChatRenderer.html) instead (which is used in [chat events](https://jd.papermc.io/paper/1.19/io/papermc/paper/event/player/AbstractChatEvent.html#renderer())). If that isn't applicable here I would ask on the Paper or Kyori discord servers to see if there are better alternatives (there probably is).

**Reply by JRoy** [reply]
Agree with pop here, individual permission checks for rgb would be impractical as people could literally change one digit and have a color that looks pretty much the same as the previous one.

**Reply by accLarson** [reply]
fair enough. Completely disregarding my hacky workaround use case above, 

If an admin wanted to grant permission to one hex color, yes a single digit change would result in an almost identical color but with the same line of thinking, the current system allows for 6 digit changes which could result in any color. 

I understand that this permission node would have minimal actual use case though.
